src/main/java/org/opentripplanner/routing/graphfinder/PlaceFinderTraverseVisitor.java

```diff
@@ -176,8 +179,22 @@ private void handleStop(RegularStop stop, double distance) {
       !seenStops.contains(stop.getId()) &&
       (filterByModes == null || stopHasPatternsWithMode(stop, filterByModes))
     ) {
-      placesFound.add(new PlaceAtDistance(stop, distance));
-      seenStops.add(stop.getId());
+      if (
+        includeStopsAndStations &&
+        stop.getParentStation() != null &&
+        !seenStops.contains(stop.getParentStation().getId()) &&
+        !seenStops.contains(stop.getId())
+      ) {
+        seenStops.add(stop.getParentStation().getId());
+        placesFound.add(new PlaceAtDistance(stop.getParentStation(), distance));
+      } else {
+        // Do not accumulate the same station many times
+        if (
+          stop.getParentStation() != null && seenStops.contains(stop.getParentStation().getId())
+        ) return;
+        seenStops.add(stop.getId());
+        placesFound.add(new PlaceAtDistance(stop, distance));
+      }
```

Comment:
I took a while for me to understand this logic. I think it would be simpler if we first check if a stop has a parent station or not and then inside those blocks do other checks etc.